Adds custom css feature. Closes #19

radiovisual opened the pull request:
Alright, this is a really big pull request!

It adds the following:
- custom css files added to `.linear/themes/` are auto-discovered by linear
- a new settings menu allows setting of the default theme
- a new settings feature lets you apply the default theme to all open rulers
- you can now launch linear in debug mode with `npm run debug` or `electron main.js --debug`
- default config files moved to version control
- default config files can be moved to the settings location if settings files are not found
- progressive installation of config files (easily "future proofed")
- a new "theme selection" menu is added to the UI

... and a whole lot more, I am sure!

Here are some screenshots of the visual changes:
### New Theme Menu

![new-theme-menu](https://cloud.githubusercontent.com/assets/5614571/14258569/d7d675a0-fa9a-11e5-85c5-677f40e2df38.png)

The new theme menu opens and closes with a nice fadein/out when the theme icon is clicked. It auto-loads all of the themes that are found in the `.linear/themes` directory.
### New Settings Menu

![new-settings-menu](https://cloud.githubusercontent.com/assets/5614571/14258575/dd357654-fa9a-11e5-89f6-cb7df3dc10c7.png)

Now you can save the theme you like the best to be the default for new rulers.

radiovisual committed:
adds custom css feature. Closes #19

radiovisual commented:
There will be a conflict if you already have a `.linear` file in your HOME directory, so in order to test this out, you will need to remove or re-name your current `.linear` file. I will write in a workaround for that feature.

radiovisual committed:
remove .linear file from older installations

radiovisual commented on src/data-store.js:
```diff
@@ -47,6 +47,12 @@ function readSettings(settingKey) {
+			// remove the .linear file from older installations of linear
+			shell.rm('-rf', settingsPath());
+		}
+
```
this will **remove** the old `.linear` file before installing the `.linear` directory with the default config file and themes, so be aware of that while testing.

radiovisual commented:
Ok, so currently, this PR will **remove** the old `.linear` file if it is found in the config location. So be aware of that while testing.

radiovisual commented:
A couple of notes:
1. The new "Apply theme to all open rulers" button in the settings menu DOES work, but there is a weird issue with electron right now that only shows the changes after you mouseover or focus the open rulers. If you merge this PR, I will open an issue to investigate it. 
2. Sometimes electron is really slow to dispatch events, so the theme updates sometimes have a slight delay. If you merge this PR, I will open up an issue to investigate.

mikaa123 commented:
OMG! That's so cool! Gonna look for it tomorrow :)

mikaa123 commented on main.js:
```diff
@@ -9,10 +9,10 @@ const Menu = electron.Menu;
+// Show debug tools in debugging mode.
+if (dataStore.debug()) {
+	require('electron-debug')({showDevTools: true});
+}
```
That's a great idea ;)

radiovisual replied:
yeah, nconf parses command line arguments for you, so I figured we could use that to our advantage. :)

mikaa123 commented on main.js:
```diff
@@ -121,6 +121,12 @@ function toggleRulerCommand() {
 app.on('ready', () => {
+	dataStore.init().then(() => {
+		setup();
+	});
```
I think we could write `dataStore.init().then(setup)`

mikaa123 commented on package.json:
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "test": "xo",
     "start": "electron main.js",
+    "debug": "electron main.js --debug",
```
:+1:

mikaa123 commented on src/app/ruler/js/ruler.js:
```diff
@@ -7,6 +7,8 @@ const ipc = require('electron').ipcRenderer;
 const Menu = remote.Menu;
 const MenuItem = remote.MenuItem;
 
+require("native-promise-only");
```
Not sure about the polyfill. Why not go directly with native promises?

radiovisual replied:
Actually, I meant to remove that, so I will remove it in my next commit.

mikaa123 commented on src/app/ruler/js/ruler.js:
```diff
@@ -43,8 +45,8 @@ const unitStrategies = {
 
 	/**
-	 * Format into to em.
-	 * @param {px} px - The number of px to convert to em.
```
woops :P

mikaa123 commented on src/data-store.js:
```diff
@@ -1,18 +1,103 @@
+		if (shell.test('-f', settingsPath())) {
+			// remove the .linear file from older installations of linear
+			shell.rm('-rf', settingsPath());
+		}
```
It's great you did this to help migration

mikaa123 commented on src/data-store.js:
```diff
@@ -1,18 +1,103 @@
+		// get a list of files (if any) at the default location
+		const settingsFiles = shell.ls(settingsPath());
+
+		if (settingsFiles.length === 0) {
```
let's be bold: `if (!settingsFiles.length)` :)

mikaa123 commented on package.json:
```diff
@@ -27,7 +28,10 @@
+    "nconf": "^0.8.2",
+    "shelljs": "^0.6.0",
+    "title-case": "^1.1.2",
+    "user-home": "^2.0.0"
```
if we choose to use it, then don't forget `native-promise-only` :)

radiovisual replied:
yeah, this is why I should have removed the references in the code files, because I had removed it from the dependencies. I think the electron runtime uses a new enough version of node that we shouldn't need to polyfill ([ponyfill?](https://ponyfoo.com/articles/polyfills-or-ponyfills#ponyfills)) Promises.

mikaa123 commented on src/defaults/config/.linear/themes/universal.css:
```diff
@@ -0,0 +1,235 @@
+    text-shadow: 1px 1px 1px rgba(119, 119, 119, 0.9);
+}
+
+.dark-theme .shdw {
```
do we still use the `.dark-theme` class?

radiovisual replied:
no, we don't need that class anymore. I should have removed all references to the `dark-theme`. Doing that now.

radiovisual committed:
tweaks

radiovisual commented:
Ok, I cleaned the code up a little bit based on your comments/suggestions. :+1:

radiovisual committed:
add customizable themes to readme

radiovisual committed:
small readme tweaks

radiovisual commented on README.md:
```diff
@@ -3,25 +3,26 @@
 
 Linear is a ruler application for Mac, made with web development in mind.
-It sits in your menu bar and doesn't get in your way. Here is what it does:
+It sits in your menu bar and doesn't get in your way. Here are a few highlights:
```
I changed `Here is what it does` to `Here are a few highlights` because the first entry on the list doesn't describe something the app _does_, rather, it describes something that the app _isn't_ (a browser extension).

radiovisual committed:
remove unnecessary hypen

mikaa123 commented on src/defaults/config/.linear/themes/dark.css:
```diff
@@ -0,0 +1,220 @@
+    }
+}
+
+.dark-theme .ruler__inner__info {
```
don't forget to remove the reference to `dark-theme` :)

radiovisual replied:
I was sure that I had removed these, because I did, so I was confused for a second, but then I realized that I only removed them from the unused `ruler.css` file and not from the `universal.css` file. Whoops! :tomato:

radiovisual replied:
But anyway, all the references to the `dark-theme` are now removed. :tada:

mikaa123 commented on src/defaults/config/.linear/themes/universal.css:
```diff
@@ -0,0 +1,225 @@
+    }
+}
+
+.dark-theme .ruler__inner__info {
```
Here too

mikaa123 commented on src/defaults/config/.linear/themes/universal.css:
```diff
@@ -0,0 +1,225 @@
+.dark-theme .ruler__inner__border,
+.dark-theme .ruler__inner__info,
+.dark-theme .ruler__inner__title {
+    border-color: black !important;
```
And there ^^

mikaa123 commented on src/defaults/config/.linear/themes/universal.css:
```diff
@@ -0,0 +1,225 @@
+}
+
+.darkCenterGuides {
+    border-color: #343434 !important;
```
Not sure if we need to keep this one?

radiovisual replied:
no, we don't need it.

mikaa123 commented:
We're almost good: there are a few remaining references to `dark-theme`

radiovisual committed:
remove unused styles

radiovisual commented:
Ok, I removed all references to the `dark-theme` in the universal.css file. I also removed the ruler.css file since it's not being used by the app anymore.

mikaa123 commented:
Merged! So cool to have this feature! Thanks

radiovisual commented:
Whoohoo! :tada:
